0000-settlement-engine/0000-settlement-engine.md

```diff
@@ -0,0 +1,353 @@
+
+1. Alice sends a custom instruction to her settlement engine to settle the fee. Note that she does not use the instruction reserved for the accounting system.
+2. Alice's settlement engine automatically negotiates the fee with Bob's settlement engine (likely subject to some limit).
+3. Alice's settlement engine sends a payment on the shared ledger or system to Bob's settlement engine for the value of the fee.
+4. Bob's settlement engine, instead of informing the accounting system of the incoming settlement, does nothing, but internally tracks that the negotiated fee was paid.
+5. If Alice's accounting system triggers any subsequent settlements to Bob, Bob's settlement engine should acknowledge them to the accounting system as normal.
+
+## Specification
+
+Operating a settlement engine requires three components each exposing a particular HTTP API: a settlement engine, an accounting system, and a transport.
+
+API endpoints exposed by these systems may be classified in two ways:
+
+1. Endpoints that may be invoked by one of these other two components, and MUST NOT be invoked manually. At a minimum, all settlement engine implementations MUST support these endpoints.
+2. Endpoints that may be invoked manually by an operator.
+
+This document outlines the first class of endpoints: those only essential to automated interoperability between these systems, and not manual operation.
+
+### Accounts and identifiers
+
+Each account MUST be identified by a unique, URL-safe string.
+
+#### Accounts vs ledger identities
+
+The settlement engine is RECOMMENDED to be responsible for correlating an account identifier to the peer's identity on the shared ledger or settlement system, if required.
+
+For separation of concerns between the accounting system and the settlement system, the accounting system is NOT RECOMMENDED to have knowledge of the peer's identity on the shared settlement system, but rather only the account identifier described here.
+
+#### Multiple accounts
+
+Settlement engine implementations SHOULD support settling with multiple accounts.
+
+#### Correlating messages
+
+The transport MUST correlate incoming messages from the peer to the correct account identifier, and MUST correlate outgoing messages from an account identifier to the correct peer to send them to.
+
+### Units and quantities
+
+#### `Quantity` type
+
+A quantity represents an amount denominated in some unit of a particular fungible asset. In uses of this type, the asset should be implicit.
+
+##### Attributes
+
+- **`amount`** string
+  - Number of the given unit of the asset between 0 and the maximum 64-bit unsigned integer, 18,446,744,073,709,551,615 (inclusive).
+  - The amount is encoded as a string to ensure no precision is lost on platforms that don't natively support full precision 64-bit integers.
+- **`scale`** number
```

Comment:
Thoughts? I can't imagine we'd possibly need anywhere close to even 255 orders of magnitude, but does u32 simplify implementation?

Reply:
Well if it's JSON we don't need to actually decide, but we could say that it must be in the range 0-255 to be valid. In theory you could have negative scales as well although everywhere we use scale we seem to require it to be positive.

Reply:
> In theory you could have negative scales as well although everywhere we use scale we seem to require it to be positive.

I think IL-DCP requires it to be unsigned?

Reply:
yeah, I just mean that conceptually a scale doesn't have to be positive. but agreed that we should not go back and change everything to support negative scale because I can't really think of a realistic use-case for it